subsys/net/l2/wifi/wifi_mgmt.c

```diff
@@ -386,6 +426,122 @@ void wifi_mgmt_raise_disconnect_result_event(struct net_if *iface, int status)
 					sizeof(struct wifi_status));
 }
 
+#ifdef CONFIG_WIFI_NM_WPA_SUPPLICANT_ROAMING
+static int wifi_start_roaming(uint32_t mgmt_request, struct net_if *iface,
+			      void *data, size_t len)
+{
+	const struct device *dev = net_if_get_device(iface);
+	const struct wifi_mgmt_ops *const wifi_mgmt_api = get_wifi_api(iface);
+
+	if (roaming_params.is_11r_used) {
+		if (wifi_mgmt_api == NULL ||
+		    wifi_mgmt_api->start_11r_roaming == NULL) {
+			return -ENOTSUP;
+		}
+
+		return wifi_mgmt_api->start_11r_roaming(dev);
+	} else if (roaming_params.is_11k_enabled) {
+		memset(&roaming_params.neighbor_rep, 0x0, sizeof(roaming_params.neighbor_rep));
+		if (wifi_mgmt_api == NULL
+		    || wifi_mgmt_api->send_11k_neighbor_request == NULL) {
+			return -ENOTSUP;
+		}
+
+		return wifi_mgmt_api->send_11k_neighbor_request(dev, NULL);
+	} else if (wifi_mgmt_api == NULL || wifi_mgmt_api->btm_query == NULL) {
+		return -ENOTSUP;
+	}
+
+	return wifi_mgmt_api->btm_query(dev, 0x10);
+}
+
+NET_MGMT_REGISTER_REQUEST_HANDLER(NET_REQUEST_WIFI_START_ROAMING, wifi_start_roaming);
+
+static int wifi_neighbor_rep_complete(uint32_t mgmt_request, struct net_if *iface,
+				      void *data, size_t len)
+{
+	const struct device *dev = net_if_get_device(iface);
+	const struct wifi_mgmt_ops *const wifi_mgmt_api = get_wifi_api(iface);
+	struct wifi_scan_params params = {0};
+
+	for (int i = 0; i < roaming_params.neighbor_rep.neighbor_cnt; i++) {
+		params.band_chan[i].channel = roaming_params.neighbor_rep.neighbor_ap[i].channel;
+		if (params.band_chan[i].channel > 14) {
+			params.band_chan[i].band = WIFI_FREQ_BAND_5_GHZ;
+		} else {
+			params.band_chan[i].band = WIFI_FREQ_BAND_2_4_GHZ;
+		}
+	}
+	if (wifi_mgmt_api == NULL || wifi_mgmt_api->candidate_scan == NULL) {
+		return -ENOTSUP;
+	}
+
+	return wifi_mgmt_api->candidate_scan(dev, &params);
+}
+
+NET_MGMT_REGISTER_REQUEST_HANDLER(NET_REQUEST_WIFI_NEIGHBOR_REP_COMPLETE,
+				  wifi_neighbor_rep_complete);
+
+void wifi_mgmt_raise_neighbor_rep_recv_event(struct net_if *iface, char *inbuf, size_t buf_len)
+{
+	const uint8_t *buf = inbuf;
+	char event[MAX_EVENT_STR_LEN] = {0};
+	char bssid[WIFI_SSID_MAX_LEN] = {0};
+	char bssid_info[WIFI_SSID_MAX_LEN]  = {0};
+	int op_class, channel, phy_type;
+	int idx = roaming_params.neighbor_rep.neighbor_cnt;
+
+	if (!buf || buf[0] == '\0') {
+		return;
+	}
+	if (sscanf(buf, "%s bssid=%s info=%s op_class=%d chan=%d phy_type=%d",
```

Comment:
Missed this in my earlier review, but this is a bit error prone as we could overwrite the buffer because there is no length for the string variables.
We could try something like (not tested)
```
if (sscanf(buf, "%" STRINGIFY(MAX_EVENT_STR_LEN) "s bssid=%" STRINGIFY(WIFI_BSSID_MAX_LEN) "s info=% " STRINGIFY(WIFI_BSSID_MAX_LEN) "s op_class=%d chan=%d phy_type=%d",
```

Reply:
As I explained before, the bssid is saved as string. If I use STRINGIFY(WIFI_BSSID_MAX_LEN), the bssid length will be incorrect.

Reply:
I would like to avoid situation where we could overwrite bssid buffer if too long bssid is given. That is why I suggested putting length to the sscanf call.
What do you mean by `bssid length will be incorrect.`, please explain?

Reply:
The BSSID is saved as string in the buffer with format "XX:XX:XX:XX:XX:XX". The string length is 17 without '\0'.
`if (sscanf(buf, "%" STRINGIFY(MAX_EVENT_STR_LEN) "s bssid=%" STRINGIFY(WIFI_BSSID_MAX_LEN) "s info=% " STRINGIFY(WIFI_BSSID_MAX_LEN) "s ...`
To my understanding, for the above code, it will be translated as below (suppose WIFI_BSSID_MAX_LEN is 6):
 `if (sscanf(buf, "%32s bssid=%6s info=%6s ...`
 For BSSID, it will only read 6 characters instead of 17, right? For event, it will read 32 characters while the actual length of event string might be less than 32. If it is true, this length check might cause errors. 
 So I think may be we can keep it for now. If I have better solutions, I can do code enhancement in the future.

Reply:
Sure, that is correct. In that case we could have here `bssid=%" STRINGIFY(WIFI_SSID_MAX_LEN)`
My point is that we should tell sscanf call what is the max size of the buffer we are working with so that it is not possible to overwrite the buffer and cause stack corruption.

Reply:
Ok, lets go with this solution at this point. Please return to this in near future.

Reply:
Yes, of course. I will try your suggestion locally. Thanks!

Reply:
@nxf58150 I created an enhancement request to fix the string length issue in sscanf, please address this asap so that we can get a fix for this in 4.0 release.